lib/url.js

```diff
@@ -75,355 +88,192 @@ const slashedProtocol = {
   return u;
 }
 
-Url.prototype.parse = function(url, parseQueryString, slashesDenoteHost) {
-  if (typeof url !== 'string') {
-    throw new TypeError("Parameter 'url' must be a string, not " + typeof url);
-  }
-
-  // Copy chrome, IE, opera backslash-handling behavior.
-  // Back slashes before the query string get converted to forward slashes
-  // See: https://code.google.com/p/chromium/issues/detail?id=25916
-  var queryIndex = url.indexOf('?'),
-      splitter =
-          (queryIndex !== -1 && queryIndex < url.indexOf('#')) ? '?' : '#',
-      uSplit = url.split(splitter),
-      slashRegex = /\\/g;
-  uSplit[0] = uSplit[0].replace(slashRegex, '/');
-  url = uSplit.join(splitter);
-
-  var rest = url;
-
-  // trim before proceeding.
-  // This is to support parse stuff like "  http://foo.com  \n"
-  rest = rest.trim();
-
-  if (!slashesDenoteHost && url.split('#').length === 1) {
-    // Try fast path regexp
-    var simplePath = simplePathPattern.exec(rest);
-    if (simplePath) {
-      this.path = rest;
-      this.href = rest;
-      this.pathname = simplePath[1];
-      if (simplePath[2]) {
-        this.search = simplePath[2];
-        if (parseQueryString) {
-          this.query = querystring.parse(this.search.substr(1));
-        } else {
-          this.query = this.search.substr(1);
-        }
-      } else if (parseQueryString) {
-        this.search = '';
-        this.query = {};
-      }
-      return this;
-    }
+Url.prototype.parse = function(str, parseQueryString, hostDenotesSlash) {
+  if (typeof str !== 'string') {
+    throw new TypeError(`Parameter 'url' must be a string, not ` +
+        typeof str);
   }
-
-  var proto = protocolPattern.exec(rest);
-  if (proto) {
-    proto = proto[0];
-    var lowerProto = proto.toLowerCase();
-    this.protocol = lowerProto;
-    rest = rest.substr(proto.length);
+  var start = 0;
+  var end = str.length - 1;
+
+  // Trim leading and trailing ws.
+  while (str.charCodeAt(start) <= 0x20 /*' '*/) start++;
+  var trimmedStart = start;
+  while (str.charCodeAt(end) <= 0x20 /*' '*/) end--;
+
+  start = this._parseProtocol(str, start, end);
+
+  // Javascript doesn't have host.
+  if (this._protocol !== 'javascript') {
+    start = this._parseHost(str, trimmedStart, start, end, hostDenotesSlash);
+    var proto = this._protocol;
+    if (!this.hostname &&
+        (this.slashes || (proto && !_slashProtocols[proto])))
+      this.hostname = this.host = '';
   }
 
-  // figure out if it's got a host
-  // user@server is *always* interpreted as a hostname, and url
-  // resolution will treat //foo/bar as host=foo,path=bar because that's
-  // how the browser resolves relative URLs.
-  if (slashesDenoteHost || proto || rest.match(/^\/\/[^@\/]+@[^@\/]+/)) {
-    var slashes = rest.substr(0, 2) === '//';
-    if (slashes && !(proto && hostlessProtocol[proto])) {
-      rest = rest.substr(2);
-      this.slashes = true;
-    }
+  if (start <= end) {
+    var ch = str.charCodeAt(start);
```

Comment:
Why get charCode and then compare to charcodes with the chars in comments, instead of just comparing the char directly?